src/find_sort_location.rs

```diff
@@ -0,0 +1,142 @@
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)> {
+        // Step1: Calculate x and 1 - x
+        let mut x_and_one_minus_x: Vec<ReplicatedSecretSharing<F>> =
+            shares.iter().map(|x| self.share_of_one - *x).collect();
+        x_and_one_minus_x.append(&mut shares.to_owned());
+
+        // Step2: Calculate cumulative sum
+        let length = x_and_one_minus_x.len();
+        let mut cumulative_sum = Vec::with_capacity(length);
+        cumulative_sum.push(x_and_one_minus_x[0]);
+        for i in 1..length {
+            cumulative_sum.push(cumulative_sum[i - 1] + x_and_one_minus_x[i]);
+        }
+
+        // Step3: Consolidate step 1 and step 2 output
+        x_and_one_minus_x
+            .iter()
+            .zip(cumulative_sum.iter())
+            .map(|(x_and_one_minus, cumulative_sum)| (*x_and_one_minus, *cumulative_sum))
+            .collect()
+    }
+
+    async fn secure_multiply(
+        &self,
+        mult_inputs: Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)>,
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        // Spawn all multiplication, wait for them to finish in parallel and then collect the results
+        let async_multiply = mult_inputs.iter().enumerate().map(|(index, input)| {
+            let output = SecureMul {
+                a_share: input.0,
+                b_share: input.1,
+                index: index as u128,
+            };
+            output.execute(self.context)
+        });
+        join_all(
+            async_multiply
+                .into_iter()
+                .map(|handle| async { handle.await.unwrap() }),
+        )
+        .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        field::Fp31,
+        helpers::{self, ring::mock::TestHelper},
+        replicated_secret_sharing::ReplicatedSecretSharing,
+    };
+
+    use super::SecureSort;
+
+    #[tokio::test]
+    async fn find_sorted_position() {
+        let ring = helpers::ring::mock::make_three();
+        let participants = crate::prss::test::make_three();
+        let context = crate::securemul::tests::make_context(&ring, &participants);
+        let mut secure_sort_output = Vec::new();
+
+        let input = [
+            (1, 0, 0),
+            (0, 0, 0),
+            (0, 1, 0),
+            (0, 0, 0),
+            (0, 0, 0),
+            (0, 0, 0),
+        ]
+        .to_vec();
+        let mut shares = [Vec::new(), Vec::new(), Vec::new()];
+
+        for iter in input {
+            let replicated_ss =
+                crate::replicated_secret_sharing::tests::secret_share(iter.0, iter.1, iter.2);
+            shares[0].push(replicated_ss.0);
+            shares[1].push(replicated_ss.1);
+            shares[2].push(replicated_ss.2);
+        }
+
+        for i in 0..3 {
+            let secure_sort = SecureSort::<TestHelper, Fp31> {
+                context: &context[i],
+                share_of_one: match i {
```

Comment:
I think that for optimizations like this, we probably want a function to do this.

Also, I would like to see this use the new HelperRole enum, so that there is no `_` arm on the match.